penquins/penquins.py

```diff
@@ -408,25 +605,153 @@ def query_skymap(
+        # we will return a dict of instances and their queries
+        if "catalogs" not in query["query"] and "catalog" not in query["query"]:
+            if name is None:
+                raise ValueError(
+                    "For catalog-less queries, please specify instance name"
+                )
+            else:
+                return {name: query}
+        # now we check if the query is for a single catalog or multiple catalogs
+        if "catalogs" in query["query"]:
+            catalogs = query["query"]["catalogs"]
+        else:
+            catalogs = [query["query"]["catalog"]]
+
+        queries = {name: None for name in self.instances.keys()}
+        if name is None:
+            # if no name is specified, but we have only one instance, we use that instance
+            if not self.multiple_instances:
+                name = list(self.instances.keys())[0]
+                if not all(
+                    [
+                        self.instance_has_catalog(catalog, name=name)
+                        for catalog in catalogs
+                    ]
+                ):
+                    raise ValueError(
+                        "One or more catalogs specified in the query are not available in the instance"
+                    )
+                queries[name] = query
+            # if no name is specified and we have multiple instances, we split the query into multiple queries by instance based on the catalogs
+            else:
+                for catalog in catalogs:
+                    instances_having_catalog = {
+                        instance_name: self.instance_has_catalog(
+                            catalog, name=instance_name
+                        )
+                        for instance_name in self.instances.keys()
+                    }
+                    if not any(instances_having_catalog.values()):
+                        raise ValueError(
+                            f"Catalog {catalog} is not available in any instance"
+                        )
+
+                    # we take the first instance that has the catalog
+                    instance_name = list(instances_having_catalog.keys())[
+                        list(instances_having_catalog.values()).index(True)
+                    ]
+                    if queries[instance_name] is None:
+                        # we set it to a copy of the original query, but with only the current catalog
+                        queries[instance_name] = deepcopy(query)
+                        # if it has a single "catalog" key, we remove it as we will replace it with a "catalogs" key
+                        if "catalog" in queries[instance_name]["query"]:
+                            del queries[instance_name]["query"]["catalog"]
+                        queries[instance_name]["query"]["catalogs"] = {
+                            catalog: query["query"]["catalogs"][catalog]
```

Comment:
Hi @Theodlz - I'm getting an error:
```
File ~/penquins/penquins/penquins.py:729, in Kowalski.prepare_query(self, query, name)
    726     if "catalog" in queries[instance_name]["query"]:
    727         del queries[instance_name]["query"]["catalog"]
    728     queries[instance_name]["query"]["catalogs"] = {
--> 729         catalog: query["query"]["catalogs"][catalog]
    730     }
    731 else:
    732     # if the instance already has a query, we add the current catalog to the list of catalogs
    733     queries[instance_name]["query"]["catalogs"][catalog] = query[
    734         "query"
    735     ]["catalogs"][catalog]

KeyError: 'catalogs'
```

when I run the following query:
```
catalog = 'Gaia_EDR3'

query = {
    "query_type": "find_one",
    "query": {
        "catalog": catalog,
        "filter": {},
    
    },
    "projection":{'_id':1},
    "kwargs":{
        "max_time_ms": 10000,
    }
}

q = kowalski.query(query)
```
where `kowalski` is initialized for all three machines.

Reply:
Hi again @bfhealy. I pushed some changes quickly (haven't tested them yet, given that the current unit tests were already passing. I need to design a test that can fail if this is wrong), maybe you can retry your query?

Reply:
@Theodlz Running the same query as before, I'm seeing the following error:
```
File ~/penquins/penquins/penquins.py:687, in Kowalski.prepare_query(self, query, name)
    685     catalogs = query["query"]["catalogs"]
    686 else:
--> 687     catalogs = {query["query"]["catalog"]: query["query"]["projection"]}
    689 queries = {name: None for name in self.instances.keys()}
    690 if name is None:
    691     # if no name is specified, but we have only one instance, we use that instance

KeyError: 'projection'
```

Reply:
Ok I will add some code later today to fix the query automatically if the user puts the projection in the wrong place. For the error below, can you retry now with the changes I just pushed?

Reply:
Thanks @Theodlz, I just ran this query:
```
catalog = 'Gaia_EDR3'

query = {
    "query_type": "find_one",
    "query": {
        "catalog": catalog,
        "filter": {},
        "projection": {'_id':1}
    },
    "kwargs":{
        "max_time_ms": 10000,
    }
}

q = kowalski.query(query)
```
and got this error:
```
File ~/penquins/penquins/penquins.py:472, in Kowalski.query(self, query, queries, name, use_batch_query, max_n_threads)
    470 else:
    471     results = {}
--> 472     for name, query in queries.items():
    473         results.update(self.single_query(query, name=name))
    474     return results

AttributeError: 'NoneType' object has no attribute 'items'
```